Add go-apidiff target

/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
- Check API compatibility between different revisions of a Go project.

Ref - https://github.com/kubernetes-sigs/controller-runtime/issues/681

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

Added `newRegion` to `api/v1beta1/ibmpowervsimage_types.go`
Renamed `Region` to `newRegion` in `api/v1alpha3/ibmvpccluster_types.go`

Sample - 

```
prajyotparab@Prajyots-MBP cluster-api-provider-ibmcloud % make apidiff                         
make -C hack/tools bin/go-apidiff
make[1]: `bin/go-apidiff' is up to date.
api/v1alpha3/ibmvpccluster_types.go
api/v1beta1/ibmpowervsimage_types.go

sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta1
  Compatible changes:
  - IBMPowerVSImageSpec.NewRegion: added

sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1alpha3
  Incompatible changes:
  - IBMVPCClusterSpec.Region: removed
  Compatible changes:
  - IBMVPCClusterSpec.NewRegion: added
make: *** [apidiff] Error 1
```

--- a/Makefile
+++ b/Makefile
@@ -31,6 +31,7 @@ TOOLS_BIN_DIR := $(TOOLS_DIR)/bin
 GO_INSTALL = ./scripts/go_install.sh
 E2E_CONF_FILE_ENVSUBST := $(REPO_ROOT)/test/e2e/config/ibmcloud-e2e-envsubst.yaml
 
+GO_APIDIFF := $(TOOLS_BIN_DIR)/go-apidiff
 GOLANGCI_LINT := $(TOOLS_BIN_DIR)/golangci-lint
 KUSTOMIZE := $(TOOLS_BIN_DIR)/kustomize
 GOJQ := $(TOOLS_BIN_DIR)/gojq
@@ -404,6 +405,20 @@ lint: $(GOLANGCI_LINT) ## Lint codebase
 .PHONY: lint-fix
 lint-fix: $(GOLANGCI_LINT) ## Lint the codebase and run auto-fixers if supported by the linter
 	GOLANGCI_LINT_EXTRA_ARGS=--fix $(MAKE) lint
+
+APIDIFF_OLD_COMMIT ?= $(shell git rev-parse origin/main)
+
+.PHONY: apidiff
+apidiff: $(GO_APIDIFF) ## Check for API differences.
+	@if ($(call checkdiff) | grep "api/"); then \
+		$(GO_APIDIFF) $(APIDIFF_OLD_COMMIT) --print-compatible; \
+	else \
+		echo "No changes to 'api/'. Nothing to do."; \
+	fi
+
+define checkdiff
+	git --no-pager diff --name-only FETCH_HEAD
+endef
 
 ALL_VERIFY_CHECKS = doctoc boilerplate shellcheck modules gen manifests conversions #tiltfile
 

--- a/hack/tools/Makefile
+++ b/hack/tools/Makefile
@@ -52,6 +52,10 @@ help:  # Display this help
 
 ##@ hack/tools:
 
+GO_APIDIFF := $(BIN_DIR)/go-apidiff
+$(GO_APIDIFF): $(BIN_DIR) go.mod go.sum ## Build a local copy of go-apidiff.
+	go build -tags=tools -o $@ github.com/joelanford/go-apidiff
+
 GOLANGCI_LINT := $(BIN_DIR)/golangci-lint
 $(GOLANGCI_LINT): $(BIN_DIR) go.mod go.sum ## Build a local copy of golangci-lint.
 	go build -tags=tools -o $@ github.com/golangci/golangci-lint/cmd/golangci-lint

--- a/scripts/ci-apidiff.sh
+++ b/scripts/ci-apidiff.sh
@@ -1,0 +1,25 @@
+#!/bin/bash
+
+# Copyright 2022 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+# shellcheck source=../hack/ensure-go.sh
+source "${REPO_ROOT}/hack/ensure-go.sh"
+
+cd "${REPO_ROOT}" && APIDIFF_OLD_COMMIT="${PULL_BASE_SHA}" make apidiff

